src/gateway.c

```diff
@@ -522,25 +522,19 @@ static void query_barrier_cb(struct barrier *barrier, int status)
 {
 	tracef("query barrier cb status:%d", status);
 	struct gateway *g = barrier->data;
-	struct handle *handle = g->req;
-	sqlite3_stmt *stmt = g->stmt;
-
-	assert(handle != NULL);
+	struct handle *req = g->req;
+	assert(req != NULL);
+	struct stmt *stmt = stmt__registry_get(&g->stmts, req->stmt_id);
 	assert(stmt != NULL);
 
-	g->stmt = NULL;
 	g->req = NULL;
 
 	if (status != 0) {
-		if (g->stmt_finalize) {
-			sqlite3_finalize(stmt);
-			g->stmt_finalize = false;
-		}
```

Comment:
I'm not entirely what this `if` block is about (I **think** it was added not to long ago to handle some corner case, probably git history will help in figuring that out), however why is it not needed anymore?

Is there a test in the pristine master code that fails when removing the above `if`? (proving that the `if` is tested and needed).

Reply:
Thanks for looking. It's unfortunate that the test passes even with the code commented out. I don't quite remember the details here, but I've seen that in follow-up commits you change things even further, so it's probably not something to worry too much about. There might be issues that were covered by this `if` that are not relevant anymore, if they are we'll hopefully trigger them in jepsen or something like that (given that the unit tests don't show it).